Auto casting of sumtype in match expression results in C error

### Describe the bug

When not inside a match expression, V is normally capable of auto-casting members of a sumtype to match a given function or method's argument type - but when inside a `match` - it results in a C compile error. See below for an MRE.

### Expected Behavior

Compile and run with output like:
```
[bug.v:34] is_button_event: true
[bug.v:34] is_button_event: true
[bug.v:29] is_button_event: true
```

### Current Behavior

```
==================
/dev/shm/v_1000/bug.767715434180678303.tmp.c:12236: error: cannot convert 'struct main__MotionEvent' to 'struct main__GameEvent'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.
```

### Reproduction Steps

```v
struct MotionEvent {}

struct ButtonEvent {}

struct WheelEvent {}

type Event = MotionEvent | ButtonEvent | WheelEvent

type GameEvent = MotionEvent | ButtonEvent

fn main() {
	be := ButtonEvent{}
	handle_event(be) // <- works, sumtype is auto-cast to correct type
	e := Event(be)
	handle_event(e) // <- works
	match e {
		MotionEvent {
			handle_game_event(e) // Ouch
		}
		ButtonEvent {
			handle_game_event(e) // Ouch
		}
		else {}
	}
}

fn handle_game_event(ge GameEvent) {
	is_button_event := ge is ButtonEvent
	dump(is_button_event)
}

fn handle_event(e Event) {
	is_button_event := e is ButtonEvent
	dump(is_button_event)
}
```

### Possible Solution

```v
struct MotionEvent {}

struct ButtonEvent {}

struct WheelEvent {}

type Event = MotionEvent | ButtonEvent | WheelEvent

type GameEvent = MotionEvent | ButtonEvent

fn main() {
	be := ButtonEvent{}
	handle_event(be) // <- works, sumtype is auto-cast to correct type
	e := Event(be)
	handle_event(e) // <- works
	match e {
		MotionEvent {
			handle_game_event(GameEvent(e)) // now, also works
		}
		ButtonEvent {
			handle_game_event(GameEvent(e)) // now, also works
		}
		else {}
	}
}

fn handle_game_event(ge GameEvent) {
	is_button_event := ge is ButtonEvent
	dump(is_button_event) // To avoid unused variable warning
}

fn handle_event(e Event) {
	is_button_event := e is ButtonEvent
	dump(is_button_event) // To avoid unused variable warning
}
```

### Additional Information/Context

_No response_

### V version

V 0.3.2 1cde554

### Environment details (OS name and version, etc.)

Ubuntu 20.04

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1718,8 +1718,8 @@ fn (mut g Gen) call_args(node ast.CallExpr) {
 				if arg.expr.obj.smartcasts.len > 0 {
 					exp_sym := g.table.sym(expected_types[i])
 					orig_sym := g.table.sym(arg.expr.obj.orig_type)
-					if orig_sym.kind != .interface_ && 
-						(exp_sym.kind != .sum_type && expected_types[i] != arg.expr.obj.orig_type) {
+					if orig_sym.kind != .interface_ && (exp_sym.kind != .sum_type
+						&& expected_types[i] != arg.expr.obj.orig_type) {
 						expected_types[i] = g.unwrap_generic(arg.expr.obj.smartcasts.last())
 						cast_sym := g.table.sym(expected_types[i])
 						if cast_sym.info is ast.Aggregate {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1718,9 +1718,8 @@ fn (mut g Gen) call_args(node ast.CallExpr) {
 				if arg.expr.obj.smartcasts.len > 0 {
 					exp_sym := g.table.sym(expected_types[i])
 					orig_sym := g.table.sym(arg.expr.obj.orig_type)
-					if orig_sym.kind != .interface_ && (exp_sym.kind != .sum_type
-						|| (exp_sym.kind == .sum_type
-						&& expected_types[i] != arg.expr.obj.orig_type)) {
+					if orig_sym.kind != .interface_ && 
+						(exp_sym.kind != .sum_type && expected_types[i] != arg.expr.obj.orig_type) {
 						expected_types[i] = g.unwrap_generic(arg.expr.obj.smartcasts.last())
 						cast_sym := g.table.sym(expected_types[i])
 						if cast_sym.info is ast.Aggregate {

--- a/vlib/v/tests/sumtype_on_match_test.v
+++ b/vlib/v/tests/sumtype_on_match_test.v
@@ -1,0 +1,35 @@
+struct MotionEvent {}
+
+struct ButtonEvent {}
+
+struct WheelEvent {}
+
+type Event = MotionEvent | ButtonEvent | WheelEvent
+
+type GameEvent = MotionEvent | ButtonEvent
+
+fn test_main() {
+	be := ButtonEvent{}
+	assert handle_event(be) == true
+	e := Event(be)
+	assert handle_event(e) == true
+	match e {
+		MotionEvent {
+			assert handle_game_event(e) == true
+		}
+		ButtonEvent {
+			assert handle_game_event(e) == true
+		}
+		else {}
+	}
+}
+
+fn handle_game_event(ge GameEvent) bool {
+	is_button_event := ge is ButtonEvent
+	return is_button_event
+}
+
+fn handle_event(e Event) bool {
+	is_button_event := e is ButtonEvent
+	return is_button_event
+}

--- a/vlib/v/tests/sumtype_on_match_test.v
+++ b/vlib/v/tests/sumtype_on_match_test.v
@@ -4,9 +4,9 @@ struct ButtonEvent {}
 
 struct WheelEvent {}
 
-type Event = MotionEvent | ButtonEvent | WheelEvent
+type Event = ButtonEvent | MotionEvent | WheelEvent
 
-type GameEvent = MotionEvent | ButtonEvent
+type GameEvent = ButtonEvent | MotionEvent
 
 fn test_main() {
 	be := ButtonEvent{}

